LOGIN_FLOW.md

```diff
@@ -0,0 +1,355 @@
+
+#### Shared state
+
+The shared state comprises all of the Chelonia state, plus a few additional keys,
+such as `loggedIn` (containing the currently logged in user).
+
+When the shared state changes, the service worker broadcasts an event to let
+each browsing context know and react to this change. For example, when a user
+joins a group, each tab needs to know of this so that the new group appears in
+the presentation of the app.
+
+#### Local state
+
+The local state comprises all of the state needed to _render a visual presentation of the app_
+(i.e., producing DOM tree and associated interactive handlers).
+
+Currently, this local state contains a copy of the shared state, but this copy
+is non-authoritative, meaning that it should only be changed when an event from
+the service worker is received to this effect.
+
+In addition, the local state contains other things that are important for
+presenting the app, such as user preferences, whether the tab is active or idle,
+etc. This information could be derived from the shared state or it could be
+entirely independent.
+
+**Crucially, this local state is not synced across browsing contexts**. When the
+local state changes, other tabs are not made aware of this fact. This allows the
+app to be open in different tabs, where each tab could be open in a different
+group or chatroom.
+
+#### Keeping the shared part of the local state in sync
+
+When the shared state changes, the service worker broadcasts an event, as
+discussed. Besides this, the service worker broadcasts a range of events to tabs
+other than simply 'the state changed'. For example, the `LOGIN` and `LOGOUT`
+events are other events that are conveyed to let individual tabs know that the
+current user session changed. In general terms, it can be said that Group Income
+(the app) is event-driven as a consequence of using a service worker architecture,
+and that presentation depends on correctly sending and processing events.
+
+## Login process
+
+Since logging in and out occurs in the browsing context as a result of user
+interaction, it's special in the sense that it's a 'shared state' event that
+originates in the browsing context. This makes the process be harder to picture
+than most other things, where there is a clearer hierarchy.
+
+In addition, there are two entry points to the log in process: one that happens
+as a result of an explicit user action (e.g., logging in) and another one that
+happens when the app is loaded in a new tab (i.e., loading Group Income in a new
+tab). The latter is so that new tabs can re-use an existing session and users
+are not prompted to log in again when they have already done so in the past.
+
+### Login events
+
+#### Overview
+
+  * `LOGIN`: The `LOGIN` event represents an active login intent, which is to
+   say, it represents that there is an active login process, which may either
+   fail or succeed. This event can originate in the service worker or in a local
+   browsing context.
```

Comment:
Could you elaborate a bit on this? Why would it originate in the service worker? Why would it originate in the browsing context? Under what circumstances?